FLT/AutomorphicRepresentation/Example.lean

```diff
@@ -482,8 +482,142 @@ lemma unitsrat_meet_unitszHat : unitsratsub ⊓ unitszHatsub = unitszsub := by
+      constructor
+      use q
+      use t
+      simp only [← Units.eq_iff, hy, ht, Units.map_mk, MonoidHom.coe_coe,
+        Algebra.TensorProduct.includeLeft_apply, Units.val_mul, one_div, q, xunit]
+      rw [← mul_one (N⁻¹ : ℚ), ← one_mul x, ← Algebra.TensorProduct.tmul_mul_tmul, div_eq_mul_inv,
+        mul_comm (v : ℚ), ← mul_one 1, ← Algebra.TensorProduct.tmul_mul_tmul, mul_assoc, mul_one]
+      congr
+      simp only [← hv, ← Units.eq_iff, Units.val_mul, Units.coe_map, MonoidHom.coe_coe,
+        Algebra.TensorProduct.includeLeft_apply, xunit, q] at wzx
+      exact wzx
+    clear * -
+    intro x hx
+    rcases canonicalForm (x⁻¹.val) with ⟨M, y, hxinv⟩
+    have : x * (i₂ y) = M := by
+      rw [← one_mul (M : QHat), ← Units.val_one, ← mul_inv_cancel x, Units.val_mul, mul_assoc]
+      congr!
+      have h : (M : QHat) = (M : ℚ) ⊗ₜ[ℤ] 1 := by norm_cast
+      rw [Algebra.TensorProduct.includeRight_apply, hxinv, h, Algebra.TensorProduct.tmul_mul_tmul,
+        mul_one, one_div, inv_mul_cancel₀]
+      simp only [ne_eq, Rat.natCast_eq_zero, PNat.ne_zero, not_false_eq_true]
+    rcases hx with ⟨X, hX⟩
+    let I := Ideal.span {X}
+    let J := (Int.castRingHom ZHat) ⁻¹' I -- this needs to be an ideal
+    have Jnonzero : (M : ℤ) ∈ J := by
+      simp only [Int.coe_castRingHom, Set.mem_preimage, Int.cast_natCast, SetLike.mem_coe, J, I]
+      rw [Ideal.mem_span_singleton']
+      use y
+      apply injective_zHat
+      simp only [mul_comm, ← hX, AddMonoidHom.coe_coe, Algebra.TensorProduct.includeRight_apply,
+        Algebra.TensorProduct.tmul_mul_tmul, mul_one, I, J] at this
+      rw [Algebra.TensorProduct.includeRight_apply, this, map_natCast]
+    let IdealJ : Ideal ℤ := by -- make this construction into a function (for Mathlib)?
+      refine ⟨⟨⟨J, ?_⟩, ?_⟩, ?_⟩
+      simp only [Int.coe_castRingHom, Set.mem_preimage, SetLike.mem_coe, Int.cast_add, J]
+      exact (fun {a b} ↦ Ideal.add_mem I)
+      simp only [Int.coe_castRingHom, Set.mem_preimage, Int.cast_zero, SetLike.mem_coe,
+        Submodule.zero_mem, J, I]
+      simp only [smul_eq_mul, I, J, Set.mem_preimage, Int.coe_castRingHom, Int.cast_mul]
+      exact (fun c {x} ↦ by apply Ideal.mul_mem_left I)
+    obtain ⟨g, hg⟩ := IsPrincipalIdealRing.principal (R := ℤ) IdealJ
+    wlog gpos : 0 < g with H
+    · specialize H x M y hxinv this X hX Jnonzero (-g)
+      apply H (by rw [← Set.neg_singleton, Submodule.span_neg, ← hg])
+      rw [Int.neg_pos, Int.lt_iff_le_and_ne, ← Int.not_gt_eq]
+      refine ⟨gpos, ?_⟩
+      rintro rfl
+      rw [Submodule.span_zero_singleton, Submodule.eq_bot_iff] at hg
+      specialize hg (M : ℤ) Jnonzero
+      simp only [Int.natCast_eq_zero, PNat.ne_zero] at hg
+    clear this hxinv y Jnonzero M
+    let N : ℕ+ := ⟨g.toNat, Int.pos_iff_toNat_pos.1 gpos⟩
+    suffices h : Ideal.span {X} = Ideal.span {(g : ZHat)} by
+      obtain ⟨y,hy⟩ : ∃ y,y*X=g := by rw [← Ideal.mem_span_singleton',h,Ideal.mem_span_singleton]
```

Comment:
```suggestion
      obtain ⟨y, hy⟩ : ∃ y, y * X = g := by 
        rw [← Ideal.mem_span_singleton',h,Ideal.mem_span_singleton]
```
Again I'm surprised there is no linter going nuts about this. Same comment for the next line too.

Reply:
Oh, this linter has only just been added to mathlib. If you merge main then a linter will be pointing out these things automatically.